### PR TITLE
[SPARK-15655] [SQL] Fix Wrong Partition Column Order when Fetching Partitioned Tables

### PR DESCRIPTION
**Pull request by gatorsmile**
#### What changes were proposed in this pull request?

When fetching the partitioned table, the output contains wrong results. The order of partition key values do not match the order of partition key columns in output schema. For example, 

``` SQL
CREATE TABLE table_with_partition(c1 string) PARTITIONED BY (p1 string,p2 string,p3 string,p4 string,p5 string)

INSERT OVERWRITE TABLE table_with_partition PARTITION (p1='a',p2='b',p3='c',p4='d',p5='e') SELECT 'blarr'

SELECT p1, p2, p3, p4, p5, c1 FROM table_with_partition
```

```
+---+---+---+---+---+-----+
| p1| p2| p3| p4| p5|   c1|
+---+---+---+---+---+-----+
|  d|  e|  c|  b|  a|blarr|
+---+---+---+---+---+-----+
```

The expected result should be 

```
+---+---+---+---+---+-----+
| p1| p2| p3| p4| p5|   c1|
+---+---+---+---+---+-----+
|  a|  b|  c|  d|  e|blarr|
+---+---+---+---+---+-----+
```

This PR is to fix this by enforcing the order matches the table partition definition.
#### How was this patch tested?

Added a test case into `SQLQuerySuite`

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @liancheng @andrewor14 @cloud-fan Could you please review this? Thanks!

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 1]
```diff
@@ -1537,6 +1537,35 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
     assert(fs.exists(path), "This is an external table, so the data should not have been dropped")
   }
 
+  test("select partitioned table") {
+    sql(
+      s"""
+         |CREATE TABLE table_with_partition(c1 string)
+         |PARTITIONED BY (p1 string,p2 string,p3 string,p4 string,p5 string)
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: I'm surprised we support this hive style syntax, cc @andrewor14

Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: There are multiple related test cases in `InsertIntoHiveTableSuite`. It has more than one bugs in this statement. For example, below is a common mistake users might make:
```
hive> CREATE TABLE partitioned (id bigint, data string) PARTITIONED BY (data string, part string);
FAILED: SemanticException [Error 10035]: Column repeated in partitioning columns
```
Currently, we return a confusing error message:
```
org.apache.spark.sql.AnalysisException: org.apache.hadoop.hive.ql.metadata.HiveException: MetaException(message:For direct MetaStore DB connections, we don't support retries at the client level.);
```

Try to submit another PR to detect these user errors and output a understandable error message.

**Comment by gatorsmile** [comment]
ping @andrewor14  : )

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
@rxin @andrewor14 @cloud-fan This returns a wrong result. Do you think we should fix it in Spark 2.0? Thanks!

**Comment by cloud-fan** [comment]
retest this please

**Comment by gatorsmile** [comment]
retest this please

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 2]
```diff
@@ -1537,6 +1537,35 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
+    // project list does not have the same order of paritioning columns in table definition
+    checkAnswer(
+      sql("SELECT p2, p3, p4, p1, p5, c1 FROM table_with_partition"),
+      Row("b", "c", "d", "a", "e", "blarr") :: Nil)
+
+    // project list contains partial partition columns in table definition
+    checkAnswer(
+      sql("SELECT p2, p1, p5, c1 FROM table_with_partition"),
+      Row("b", "a", "e", "blarr") :: Nil)
+  }
```
Comment (clockfly) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Should we drop the table after test?

Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: I think we can use `withTable` at the beginning

**Comment by clockfly** [comment]
Looks good to me except one minor test issue.

**Comment by gatorsmile** [comment]
@cloud-fan @clockfly Thank you for your review! Let me know if the latest code changes look fine. Thanks!

**Comment by cloud-fan** [comment]
LGTM, pending jenkins

**Comment by cloud-fan** [comment]
thanks, merging to master/2.0!
